haskell/indexer.bzl

```diff
@@ -0,0 +1,168 @@
+    log_file = ctx.actions.declare_file(_get_named_path("kythe.log", unique_id))
+
+    args = ctx.actions.args()
+
+    compile_flags = ctx.actions.args()
+    compile_flags.add_all(target[HaskellInfo].compile_flags)
+    compile_flags.add_all(target[HaskellInfo].source_files)
+
+    locale_archive_depset = (
+        depset([hs.toolchain.locale_archive]) if hs.toolchain.locale_archive != None else depset()
+    )
+
+    # C library dependencies for runtime.
+    (ghci_extra_libs, ghc_env) = get_ghci_extra_libs(
+        hs,
+        posix,
+        target[CcInfo],
+        # TODO(robinp): can drop?
+        # haddock changes directory during its execution. We prefix
+        # LD_LIBRARY_PATH with the current working directory on wrapper script
+        # startup.
+        path_prefix = "$PWD",
+    )
+
+    indexer_wrapper = ctx.actions.declare_file("indexer_wrapper-{}".format(hs.name))
+    ctx.actions.expand_template(
+        template = ctx.file._indexer_wrapper_tpl,
+        output = indexer_wrapper,
+        substitutions = {
+            "%{indexer}": ctx.file._indexer_binary.path,
+            "%{outfile}": entries_file.path,
+            "%{logfile}": log_file.path,
+            # XXX Workaround
+            # https://github.com/bazelbuild/bazel/issues/5980.
+            "%{env}": render_env(dicts.add(hs.env, ghc_env)),
+        },
+        is_executable = True,
+    )
+
+    ctx.actions.run(
+        inputs = depset(transitive = [
+            target[HaskellInfo].package_databases,
+            target[HaskellInfo].interface_dirs,
+            target[HaskellInfo].source_files,
+            target[HaskellInfo].extra_source_files,
+            target[HaskellInfo].dynamic_libraries,
+
+            # Seems needed for TH magiclink - why aren't the dynamic libs used
+            # there?
+            target[HaskellInfo].static_libraries,
+
+            target[HaskellInfo].interface_dirs, # ??
+            ghci_extra_libs,
+            target[CcInfo].compilation_context.headers,
+            depset([ctx.file._indexer_binary]),
+            depset([
+                hs.tools.ghc_pkg,
+            ]),
+            locale_archive_depset,
+        ]),
+        tools = [
+          hs.tools.ghc,  # To get libdir
+        ],
+        outputs = [entries_file, log_file],
+        mnemonic = "HaskellIndexer",
+        progress_message = "HaskellIndexer {}".format(ctx.label),
+        executable = indexer_wrapper,
+        arguments = [
+            args,
+            compile_flags,
+        ],
+        env = {
+            "PATH": (";" if hs.toolchain.is_windows else ":").join(posix.paths),
+            "GHC": hs.tools.ghc.path,
+            "CORPUS": ctx.workspace_name,
+        },
+    )
+
+    indexer_info = HaskellIndexerInfo(
+        entries = entries_file,
+    )
+    output_files = OutputGroupInfo(default = [entries_file, log_file])
+
+    return [indexer_info, output_files]
+
+haskell_indexer_aspect = aspect(
```

Comment:
I think we'll also need some tests for this. Similar aspects/rules like haddock, docstring, or REPL, often break when we make changes that touch compiler or linker flags or inputs.